### PR TITLE
Add api stripe checkout integration

### PR DESCRIPTION
**Pull request by ahmaxed**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

### CODE REVIEW

**Comment by scissorsneedfoodtoo** [comment]
Tested this locally with several test cards on the donate page and after completing a certification. The donations all show up on the Stripe's Test Data dashboard, and the redirect after donating on the donation page works as expected.

Everything LGTM :+1: :+1:

**Review by scissorsneedfoodtoo — DISMISSED** [review]

File: client/src/components/Donation/DonateForm.js [hunk 1]
```diff
@@ -430,29 +420,32 @@ class DonateForm extends Component {
 
   render() {
     const {
-      donationFormState: { processing, success, error },
+      donationFormState: { processing, success, error, redirecting },
       isMinimalForm
     } = this.props;
+    console.log({ processing, success, error, redirecting });
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/DonateForm.js:426]:
```suggestion
```

File: client/src/redux/donation-saga.js [hunk 2]
```diff
@@ -55,6 +59,26 @@ function* addDonationSaga({ payload }) {
...
+      sessionId: session.data.id
+    });
+  } catch (error) {
+    const data =
+      error.response && error.response.data
+        ? error.response.data
+        : {
+            message: defaultDonationError
+          };
+    yield put(addDonationError(data.message));
```
Comment (ojeytonwilliams) [inline client/src/redux/donation-saga.js:78]:
```suggestion
    const err =
      error.response && error.response.data
        ? error.response.data.message
        : defaultDonationError;
    yield put(addDonationError(err));
```
Saves a couple of lines.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api-server/src/server/boot/donate.js [hunk 3]
```diff
@@ -246,6 +250,51 @@ export default function donateBoot(app, done) {
...
+        line_items: [
+          {
+            price,
+            quantity: 1
+          }
+        ],
+        metadata: { ...body },
+        mode: isSubscription ? 'subscription' : 'payment',
+        success_url: donationUrls.successUrl,
+        cancel_url: donationUrls.cancelUrl
```
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply api-server/src/server/boot/donate.js:285]: We don't need to, they will come back to the client they came from :)

[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api-server/src/server/boot/donate.js:287]: Not for the /news redirect, surely?

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Comment by ojeytonwilliams** [comment]
@scissorsneedfoodtoo could you give this a final approve if you're happy?  We lost your first review when we added some commits.

File: api-server/src/server/boot/donate.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply api-server/src/server/boot/donate.js:287]: So when creating the session in our api, we will pass the localized client url from which the button was clicked. 
When the donation is complete, stripe will use that redirect url and send users back to the desired client.

**Comment by scissorsneedfoodtoo** [comment]
Thanks for your patience, and also thanks for catching those things @ojeytonwilliams.

I was having some trouble testing this because it occasionally it would say that the page at `/donate/create-stripe-session` didn't exist.

It may have been an issue with my cache or keys, though. After checking the keys and doing a hard refresh everything seems to be fine.

File: api-server/src/server/boot/donate.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply api-server/src/server/boot/donate.js:287]: I just tested this in other locales and can confirm that we only redirect to the English version of the news or back to the donate page on success or failure.

It seems like we could add something like this to `config/donation-settings.js`, or somewhere else in the config directory to get localized redirects:

```js
const envData = require('./env.json');

const { clientLocale } = envData;

...

const donationUrls = {
  espanol: {
    successUrl:
      'https://www.freecodecamp.org/espanol/news/thank-you-for-donating/',
    cancelUrl: 'https://www.freecodecamp.org/espanol/donate/'
  },
  chinese: {
    successUrl: 'https://chinese.freecodecamp.org/news/thank-you-for-donating/',
    cancelUrl: 'https://chinese.freecodecamp.org/donate/'
  },
  english: {
    successUrl: 'https://www.freecodecamp.org/news/thank-you-for-donating/',
    cancelUrl: 'https://www.freecodecamp.org/donate/'
  }
}[clientLocale];

...
```

But we don't have "Thank You for Donating" pages set up for the other locales yet, so I'm fine with putting this off for a bit until we create those.

**Review by scissorsneedfoodtoo — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
Ahmad and I ran into that too, when testing it out. You need all the donation keys to be present for that page to get created. There's a warning, but it's not obvious (we missed it).

**Comment by scissorsneedfoodtoo** [comment]
Awesome, thanks for confirming that. Good to know it wasn't just me. I remember having trouble until I added the Stripe and PayPal keys back in. They were already present, but I might have accidentally changed one in the `.env` file without noticing.

**Pull request merged**